src/fermilib/utils/_parallel_bubble_sort.py

```diff
@@ -0,0 +1,114 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"""This module has functions to perform parallel bubble sort on arrays."""
+from functools import partial
+from itertools import product
+
+
+def index(coordinate_ordering, system_side_length, position):
+    """Return the index of an n-dimensional position in an 1-dimensional
+    array.
+
+    The index is computed using the given ordering of coordinate
+    weights as well as the side length of the system from which the
+    n-dimensional position is taken.
+
+    Args:
+        position: The vector (list) of integers indicating the position
+                  within the n-dimensional system.
+        system_side_length: The number of points along a side of the system.
+        coordinate_ordering: Tuple indicating the order of coordinates
+                             by which to increment the index.
+
+    Returns:
+        The integer index of the position in a 1D array.
+
+    Examples:
+        index((2, 1, 0), 4, (1, 2, 3)) -> 27 (16 + 4 * 2 + 1 * 3).
+        This is standard lexicographic ordering for a 3D array.
+
+        index((0, 1, 2), 2, (0, 1, 1)) -> 6 (2 * 1 + 4 * 1)
+
+        index((2, 0, 1), 3, (2, 1, 0)) -> 19 (9 * 2 + 1 + 3 * 0)
+    """
+    return sum(position[i] * system_side_length ** coordinate_ordering[i]
+               for i in range(len(coordinate_ordering)))
+
+
+def is_sorted_embedded_array(arr, key, system_side_length):
+    """Determine whether an embedded n-dim array of positions is sorted.
+
+    Args:
+        arr: the array to work with.
+        key: the key function to use to determine ordering.
+        system_side_length: The integer side length of the n-dimensional
+                            cube from which the positions in the array
```

Comment:
I think in most of our code we use the convention of two indents after a breakline in args.